Documentation about C++ exceptions may leads to errors

Hello,

Just a little message about how C++ exceptions handling is described in the documentation especially when targeting JavaScript : the documentation is explaining us to use something like this : 
```
%exception {
  try {                                                     
    $action
  } catch (const std::runtime_error &e) {
    SWIG_exception(SWIG_RuntimeError, e.what());
  } catch (...) {
    SWIG_exception(SWIG_RuntimeError, "unknown exception");
  } 
} 
``` 

It's working for most of the situations but when an error is thrown in the constructor of an object, it makes the app crashing (by calling the destructor while the object wasn't created). I found a solution by adding a `goto fail;`after the calls to `SWIG_exception`which seems to work actually.

Maybe the documentation needs an update to deal with this situation :wink:.


Thanks for making this nice tool which saved a lot of my time, I hope this helps =).

26bbc96d58b81293a2357ba3ba7c4419d87417dd added a new `swig_exception` testcase for this, which gives these warnings with Javascript because `SWIG_exception` doesn't return there:

```
checking javascript (node) testcase swig_exception
../../swig_exception_wrap.cxx: In function ‘SwigV8ReturnValue _wrap_new_Square(const SwigV8Arguments&)’:
../../swig_exception_wrap.cxx:2216:75: error: ‘result’ may be used uninitialized in this function [-Werror=uninitialized]
../../swig_exception_wrap.cxx: In function ‘SwigV8ReturnValue _wrap_new_Circle(const SwigV8Arguments&)’:
../../swig_exception_wrap.cxx:2080:75: error: ‘result’ may be used uninitialized in this function [-Werror=uninitialized]
../../swig_exception_wrap.cxx: In function ‘SwigV8ReturnValue _wrap_new_Value(const SwigV8Arguments&)’:
../../swig_exception_wrap.cxx:1664:74: error: ‘result’ may be used uninitialized in this function [-Werror=uninitialized]
../../swig_exception_wrap.cxx: In function ‘SwigV8ReturnValue _wrap_Square_perimeter(const SwigV8Arguments&)’:
../../swig_exception_wrap.cxx:2282:48: error: ‘result’ may be used uninitialized in this function [-Werror=uninitialized]
../../swig_exception_wrap.cxx: In function ‘SwigV8ReturnValue _wrap_Square_area(const SwigV8Arguments&)’:
../../swig_exception_wrap.cxx:2248:48: error: ‘result’ may be used uninitialized in this function [-Werror=uninitialized]
../../swig_exception_wrap.cxx: In function ‘SwigV8ReturnValue _wrap_Circle_perimeter(const SwigV8Arguments&)’:
../../swig_exception_wrap.cxx:2146:48: error: ‘result’ may be used uninitialized in this function [-Werror=uninitialized]
../../swig_exception_wrap.cxx: In function ‘SwigV8ReturnValue _wrap_Circle_area(const SwigV8Arguments&)’:
../../swig_exception_wrap.cxx:2112:48: error: ‘result’ may be used uninitialized in this function [-Werror=uninitialized]
../../swig_exception_wrap.cxx: In function ‘SwigV8ReturnValue _wrap_Shape_perimeter(const SwigV8Arguments&)’:
../../swig_exception_wrap.cxx:1963:48: error: ‘result’ may be used uninitialized in this function [-Werror=uninitialized]
../../swig_exception_wrap.cxx: In function ‘SwigV8ReturnValue _wrap_Shape_area(const SwigV8Arguments&)’:
../../swig_exception_wrap.cxx:1929:48: error: ‘result’ may be used uninitialized in this function [-Werror=uninitialized]
```

--- a/CHANGES.current
+++ b/CHANGES.current
@@ -6,6 +6,10 @@ Version 3.0.11 (in progress)
 ============================
 
 2016-09-16: olly
+            [Javascript] Fix SWIG_exception() macro to return from the current
+            function.  Fixes #789, reported by Julien Dutriaux.
+
+2016-09-16: olly
             [PHP] Fix SWIG_exception() macro to return from the current function.
             Fixes #240, reported by Sergey Seroshtan.
 

--- a/Examples/test-suite/javascript/swig_exception_runme.js
+++ b/Examples/test-suite/javascript/swig_exception_runme.js
@@ -1,0 +1,30 @@
+var swig_exception = require("swig_exception");
+
+var c = new swig_exception.Circle(10);
+var s = new swig_exception.Square(10);
+
+if (swig_exception.Shape.nshapes != 2) {
+    throw "Shape.nshapes should be 2, actually " + swig_exception.Shape.nshapes;
+}
+
+// ----- Throw exception -----
+try {
+    c.throwException();
+    throw "Exception wasn't thrown";
+} catch (e) {
+    if (e.message != "OK") {
+	throw "Exception message should be \"OK\", actually \"" + e.message + "\"";
+    }
+}
+
+// ----- Delete everything -----
+
+c = null;
+s = null;
+e = null;
+
+/* FIXME: Garbage collection needs to happen before this check will work.
+if (swig_exception.Shape.nshapes != 0) {
+    throw "Shape.nshapes should be 0, actually " + swig_exception.Shape.nshapes;
+}
+*/

--- a/Lib/javascript/jsc/javascriptcode.swg
+++ b/Lib/javascript/jsc/javascriptcode.swg
@@ -32,6 +32,7 @@ static JSObjectRef $jswrapper(JSContextRef context, JSObjectRef ctorObject,
     size_t argc, const JSValueRef argv[], JSValueRef* exception)
 {
     SWIG_exception(SWIG_ERROR, "Class $jsname can not be instantiated");
+fail:
     return 0;
 }
 %}

--- a/Lib/javascript/jsc/javascripthelpers.swg
+++ b/Lib/javascript/jsc/javascripthelpers.swg
@@ -53,7 +53,7 @@ SWIGINTERN bool JS_veto_set_variable(JSContextRef context, JSObjectRef thisObjec
     } else {
       SWIG_exception(SWIG_ERROR, msg);
     }
-    
+fail:    
     return false;
 }
 

--- a/Lib/javascript/jsc/javascriptrun.swg
+++ b/Lib/javascript/jsc/javascriptrun.swg
@@ -4,7 +4,7 @@
  * ---------------------------------------------------------------------------*/
 
 #define SWIG_Error(code, msg)     SWIG_JSC_exception(context, exception, code, msg)
-#define SWIG_exception(code, msg) SWIG_JSC_exception(context, exception, code, msg)
+#define SWIG_exception(code, msg) do { SWIG_JSC_exception(context, exception, code, msg); SWIG_fail; } while (0)
 #define SWIG_fail                 goto fail
 
 SWIGRUNTIME void SWIG_Javascript_Raise(JSContextRef context, JSValueRef *exception, const char* type) {

--- a/Lib/javascript/v8/javascriptcode.swg
+++ b/Lib/javascript/v8/javascriptcode.swg
@@ -36,6 +36,7 @@ static SwigV8ReturnValue $jswrapper(const SwigV8Arguments &args) {
   SWIGV8_HANDLESCOPE();
   
   SWIG_exception(SWIG_ERROR, "Class $jsname can not be instantiated");
+fail:
   SWIGV8_RETURN(SWIGV8_UNDEFINED());
 }
 %}

--- a/Lib/javascript/v8/javascripthelpers.swg
+++ b/Lib/javascript/v8/javascripthelpers.swg
@@ -83,6 +83,7 @@ SWIGRUNTIME void JS_veto_set_variable(v8::Local<v8::String> property, v8::Local<
     } else {
       SWIG_exception(SWIG_ERROR, msg);
     }
+fail: ;
 }
 
 %} // v8_helper_functions

--- a/Lib/javascript/v8/javascriptrun.swg
+++ b/Lib/javascript/v8/javascriptrun.swg
@@ -97,7 +97,7 @@ typedef v8::PropertyCallbackInfo<v8::Value> SwigV8PropertyCallbackInfo;
  * ---------------------------------------------------------------------------*/
 
 #define SWIG_Error(code, msg)     SWIGV8_ErrorHandler.error(code, msg)
-#define SWIG_exception(code, msg) SWIGV8_ErrorHandler.error(code, msg)
+#define SWIG_exception(code, msg) do { SWIGV8_ErrorHandler.error(code, msg); SWIG_fail; } while (0)
 #define SWIG_fail                 goto fail
 #define SWIGV8_OVERLOAD false
 

